source/NRF52Serial.cpp

```diff
@@ -1,139 +1,349 @@
 #include "NRF52Serial.h"
-
-volatile int tx_complete = 1;
-volatile int rx_complete = 1;
+#include "peripheral_alloc.h"
+#include "NotifyEvents.h"
 
 using namespace codal;
 
-void nrf_uarte_handler(nrfx_uarte_event_t const * p_event, void * p_context)
+extern int8_t target_get_irq_disabled();
+
+/**
+ * Constructor
+ *
+ * @param tx the pin instance to use for transmission
+ *
+ * @param rx the pin instance to use for reception
+ *
+ **/
+NRF52Serial::NRF52Serial(Pin& tx, Pin& rx, NRF_UARTE_Type* device) 
+ : Serial(tx, rx), is_configured_(false), is_tx_in_progress_(false)
+{
+    if(device != NULL)
+        p_uarte_ = (NRF_UARTE_Type*)allocate_peripheral((void*)device);
+    configurePins(tx,rx);
+}
+
+NRF52Serial::~NRF52Serial()
 {
-    if (p_context == NULL)
-        return;
+    nrf_uarte_int_disable(p_uarte_, NRF_UARTE_INT_RXDRDY_MASK|
+                                    NRF_UARTE_INT_ENDRX_MASK |
+                                    NRF_UARTE_INT_ENDTX_MASK |
+                                    NRF_UARTE_INT_ERROR_MASK |
+                                    NRF_UARTE_INT_RXTO_MASK  |
+                                    NRF_UARTE_INT_TXSTOPPED_MASK);
+    NVIC_DisableIRQ(get_alloc_peri_irqn(p_uarte_));
 
-    NRF52Serial* instance = (NRF52Serial*) p_context;
+    // Make sure all transfers are finished before UARTE is disabled
+    // to achieve the lowest power consumption.
+    nrf_uarte_shorts_disable(p_uarte_, NRF_UARTE_SHORT_ENDRX_STARTRX);
+    nrf_uarte_task_trigger(p_uarte_, NRF_UARTE_TASK_STOPRX);
+    nrf_uarte_event_clear(p_uarte_, NRF_UARTE_EVENT_TXSTOPPED);
+    nrf_uarte_task_trigger(p_uarte_, NRF_UARTE_TASK_STOPTX);
+    while (!nrf_uarte_event_check(p_uarte_, NRF_UARTE_EVENT_TXSTOPPED))
+    {}
 
-    if (p_event->type == NRFX_UARTE_EVT_TX_DONE)
-    {
-        tx_complete = 1;
-        instance->dataTransmitted();
+    nrf_uarte_disable(p_uarte_);
+    nrf_uarte_txrx_pins_disconnect(p_uarte_);
```

Comment:
Free serial instance?

Reply:
Done on 9f6a6ce.

Reply:
This is necessary for SPI and I2C, but since this PR is about Serial, I will PR about each later.